### DISCUSSION TITLE
Mode 2031 (Theme Notifications) update coalescing

### DISCUSSION THREAD

**Discussion by gpanders** [contour-terminal/contour#1657] [Development]
Hello!

I'm currently implementing support for theme notifications (Mode 2031) into Neovim based on the specification here: https://github.com/contour-terminal/contour/blob/f3c3334aa5c861348c5bbe8ffe572c872eef2e08/docs/vt-extensions/color-palette-update-notifications.md

The specification requires sending a notification each time the terminal's color palette changes. However, the spec is ambiguous on whether or how these notifications should be coalesced. For instance, if the terminal's foreground, background, and cursor color are all changed simultaneously, should the terminal send 3 notifications or just 1?

The terminal should _probably_ just send a single notification (I can't think of any reason why sending multiple would be necessary), but since this is not explicitly stated in the specification, terminals aren't required to do this. The result is that some terminals may send many redundant notifications which applications must deal with. The problem is even more exacerbated when one considers individual palette color updates. If an application updates, say, 20 palette colors at once (using OSC 4), the terminal should send a single notification, not 20.

If you agree, can we update the specification to make this behavior required?

**Comment by Yaraslaut** [comment]
Hi @gpanders, thanks for the interest in this specs, we will adjust specification to point out that terminal should merge notifications to not spam application to much, and also a bit from  @lhecker "I'd be happy to add support for it in windows terminal. it would probably be available in 1.24 and later"

**Comment by christianparpart** [comment]
root cause found and tracker ticket on our side is closed (#1659). The actual changes required are being tracked here: https://github.com/ghostty-org/ghostty/issues/3965
